Inline Chat: Separate from sidebar (and ChatProvider)

umpox opened the pull request:
## Description

This PR:
- Decouples the sidebar and inline chat interfaces from the messaging logic in `ChatViewProvider`. 
  - Moved shared context gathering (e.g. files/selection/workspace changed events) to `ContextProvider`.
  - Moved core messaging logic (e.g. building prompts, sending and receiving messages, saving and restoring history). to `MessageProvider`
  - Added `SidebarChatViewProvider` and `InlineChatViewProvider` to handle creating derived instances of `MessageProvider` and displaying Cody's responses to both the sidebar and inline chat windows.
- Updates inline chat to be independent but still have close parity to the sidebar chat
  - Makes each inline chat window have its own isolated context.
    - Messages are no longer shared between inline chats and the sidebar
    - Messages are no longer shared between multiple inline chats
  - Added stop generating button for inline chats.
  - Removed the "loading" icon, as it was duplicative of the new inline chat "..." response - and misleading when there was more than one inline chat.
  - Added CTA to move a chat from the inline window into the sidebar, allowing users to take a file discussion into something more general in the sidebar.


## Examples

**Async chats with the sidebar, and moving chats from inline -> sidebar**

https://github.com/sourcegraph/cody/assets/9516420/81e7199b-cc16-4365-88a8-d7ce84ee2f97

**Stop generating for inline chat**

https://github.com/sourcegraph/cody/assets/9516420/59135259-5e26-4d7e-a66b-3e46e2c654e5


## Test plan

Tested in VS Code:
- Inline chat (+ stop generating)
- Inline fix
- Inline touch
- Sidebar chat
- Transferring inline to sidebar

umpox commented on vscode/src/services/InlineController.ts:
```diff
@@ -437,14 +462,24 @@ class Comment implements vscode.Comment {
          * Relevant VS Code logic: https://sourcegraph.com/github.com/microsoft/vscode@6c8cdf325eb1dc8a0e2ea9205a1d2ca05f69c101/-/blob/src/vs/workbench/api/common/extHostComments.ts?L461-492
          */
-        this.update = throttle(this.update.bind(this), 100)
+        this.update = throttle(this.unthrottledUpdate.bind(this), 500)
```
You can see an example of the flicking in the 2nd demo video above, it's frustrating and still does cause some misclicks. I'll raise an issue on the VS Code GitHub, we might not be able to get a fix for this unless we put work in ourselves (as it's probably an unusual case for streaming comments). Debatable if it's worth doing that right now given this might not be the ideal long term inline UX (interactive api would be ideal).

I feel the benefits of streaming responses still outweigh the flickering, but happy to discuss

abeatrix reviewed (commented):
Clicking on the default recipes no longer take me back to the chat view, is that intended?

abeatrix reviewed (changes requested):
Thanks for making such a great improvement! I love the stop button and the move chat to sidebar button you added for Inline Chat, those are awesome!

I did encounter an issue: The "Stop Generating" button would stay indefinitely after Cody has replied.  Swtiching to a different instance does not help:
![image](https://github.com/sourcegraph/cody/assets/68532117/89cf728b-6eed-4b0c-8f74-1f10ae2d249f)
Loom: https://www.loom.com/share/07d46c653ba04d6fa5da1d2cd6d00618?sid=f7d1771e-b891-4c2c-852e-417900761ec7

Update: Can't reproduce it again after reloading the window! Clicking on the recipes are still not redirecting me back to chat view though, thought the recipe wasn't working at all 😅

abeatrix commented:
Regarding issue 2, it's weird because the UI would not display the new comment thread, but if i input /fix comment in the comment thread, the selected would still get fixed, just without the comment thread and no code lenses and decoration as well

I'm also having issue with Inline Fixups where the return code is replace with the wrong range. it happens twice for me but i don't know what I did to provide reproduce instructions though:
![buginlinefixups](https://github.com/sourcegraph/cody/assets/68532117/1f31523e-1c2b-43a3-9e8c-fd9fd7917233)

The Undo button doesn't work as expected when the aforementioned issue happens:
![inlineundo](https://github.com/sourcegraph/cody/assets/68532117/e18b5e64-ebf6-48fd-80d7-216d7f6a654b)

umpox commented:
@abeatrix Thanks for testing!

> Clicking on the recipes are still not redirecting me back to chat view though, thought the recipe wasn't working at all

Think I know why this is, thought I fixed it but guess not.

I haven't been able to replicate the "Stop generating" and "Ask Cody" button issues, I think they both related to the `cody.reply.pending` state 🤔  I refactored that to be `cody.inline.reply.pending` so wondering if there's a weird issue where context isn't updating correctly. I'll investigate.

umpox commented:
@abeatrix I think there is something weird going on with fixups as I noticed that same issue with the replace range being wrong on `main` 🤔 

I did replicate it on my branch once, and then couldn't! Not sure what is causing it - really weird. Going to look into improving the /fix output after this PR so maybe I can find out there.

There is MAYBE a reason this happened here... which could be because I was setting a key for the threads based off position in the doc (forgetting that people actually write code which moves other code in the doc 🤦 ). I've fixed that now!

abeatrix commented:
#### tested

- ✅  recipes
- ✅  inline chat
- ✅  stop gen
- 🟡  transferring to sidebar
- ⛔ inline fix
- 🟡  inline touch
- [ ] non-stop fix
- [ ] custom recipes
- [ ] plugins

Still have a few things to test but so far so good! I think most of the issues I have encountered are not big problems (the ones marked as  🟡) except for Inline fix, which I think we should look into next (separately from this PR) as the CEs use it for demos occasionally. 

For more context: I tried downgrading the extension to 0.4.2 and the inline fixups work fine there, so I looked at some of the changes made in 0.4.3, and it seems like reverting [the cb function in this change ](https://github.com/sourcegraph/sourcegraph/pull/54522/files#diff-c0c0099a5c7ffbf8ed718d1268daaeefa38349d2c8c2f7cb41721dd99d9aa90a) makes the issue go away, do you think that is the cause here?

#### encountered issues 

⛔ Inline Fix

The code  is unusable and wrapped with weird tags 👀 
<img width="1407" alt="image" src="https://github.com/sourcegraph/cody/assets/68532117/ed941ce4-6acb-4546-8a9f-dbc59d62151f">

🟡  Transferring to sidebar: 

once I've transferred the conversation to the sidebar, and then continued talking to Cody in inline Chat, and then clicked the transfer to sidebar button again, the new replies were not added to the sidebar. Maybe we could delete the inline chat automatically when the current session has been moved to the sidebar?
<img width="1879" alt="image" src="https://github.com/sourcegraph/cody/assets/68532117/d7d70ce8-5e03-4c9b-a302-bbc5c3bd7fa4">

🟡  inline touch

the returned code in the new file is wrapped in markdown format 🤔 

<img width="1564" alt="image" src="https://github.com/sourcegraph/cody/assets/68532117/e8ac6a3a-f4f5-47be-b342-c7908fb795f0">

umpox commented:
Updated with `main` and tested:
- ✅ Sidebar chat + recipes
- ✅ Inline chat + stop gen + moving to sidebar. I have made the inline chat delete itself when moving to the sidebar.
- ✅ Inline fix
- ✅ Non stop fix
- ✅ Custom recipes
- [ ] Plugins


⛔ Inline touch - Seems to be an issue with inline touch on `main`. Going to investigate now, unrelated to this branch so shouldn't block merging.

@abeatrix The issues you saw with inline fix and inline touch are highly unlikely to be related to this change. I see those same issues on `main`. The recipe I want to look at is the "fix" flow, so I'm looking into this next. Our current work often leads to these issues but I have a few ideas on how we can improve it, will share more on Slack today.

FWIW I can also replicate these issues before the change to introduce the typewriter effect. That also shouldn't impact this at all as it just changes the response chunk size - we don't make any assumptions about response chunks AFAIK (e.g. expecting three backticks to come in at once), if we did I think we would have a lot of other ideas. I think it's more of an issue with the prompt we're using (e.g. markdown tags being included even though we ask it not to, poor response quality)

abeatrix commented:
#### tested

- ✅  recipes
- ✅  inline chat
- ✅  stop gen
- 🟡  transferring to sidebar
- ⛔ inline fix
- 🟡  inline touch
- [ ] non-stop fix
- ⛔ custom recipes
- ✅  Auth - switch instances
- ✅  Auth - sign in sign out
- [ ] plugins

# Issue with Custom Recipes

The button is not working as it currently is in main:

![Screenshot 2023-07-24 at 8 20 04 AM](https://github.com/sourcegraph/cody/assets/68532117/a09fc7c4-6cf6-4924-8c0a-275571b185cc)
![customrecipesmenubug](https://github.com/sourcegraph/cody/assets/68532117/db1cabb3-8ff1-41ce-979c-1c575552f953)

In main, the Custom Recipes menu would stay instead of redirecting the user back to the Chat view:
![image](https://github.com/sourcegraph/cody/assets/68532117/644de3b1-7b58-451c-8cca-1b736e20ad2f)

abeatrix commented:
@umpox im running into the same issue with the recipes menu after i have make a selection in the menu. It would move back to the chat view instead of staying in the recipes view 👀

abeatrix reviewed (approved):
Everything seems to be working for me now, will need some follow-ups for:
- Clicking on the `stop generating` button before LLM response would not cancel the request even when the button has disappeared
- Inline Fixup is returning unusable code (not due to changes in this PR)

Thank you so much for clearing the repo for us!

umpox commented:
Updated with `main`, ran some final checks and everything looks good. Will merge now!
